lib/wherebot.rb

```diff
@@ -0,0 +1,142 @@
+require 'net/imap'
+require 'mail'
+
+class Wherebot
+  TRASH = "Trash".freeze
+  INBOX = "eff-where".freeze
+
+  class << self
+    def update_wheres(destructive: false)
+      imap = Net::IMAP.new(
+        ENV['INCOMING_MAIL_SERVER'], ENV['INCOMING_MAIL_PORT'], true
+      )
+
+      sign_into_wheremail(imap)
+
+      new_messages(imap).each do |id|
+        Wherebot::Message.new(imap, id, destructive).create
+      end
+
+      clean_up(imap, destructive)
+    end
+
+    private
+
+    def sign_into_wheremail(imap)
+      imap.login(ENV['WHEREBOT_USER_NAME'], ENV['WHEREBOT_PASSWORD'])
+      imap.examine(INBOX)
+    end
+
+    def new_messages(imap)
+      @new_messages ||= imap.search(["UNSEEN"])
+    end
+
+    def clean_up(imap, destructive)
+      imap.expunge if destructive
+      imap.logout
+      imap.disconnect
+    end
+  end
+
+  class Message
+    def initialize(imap, email_id, destroy = false)
+      @imap = imap
+      @id = email_id
+      @destroy = destroy
+    end
+
+    def create
+      wm = WhereMessage.find_or_create_by(
+        provenance: 'where@eff.org',
+        sent_at: DateTime.parse(headers.date),
+        user: User.find_by(
+          email: "#{headers.from[0].mailbox}@#{headers.from[0].host}"
+        )
+      )
+      if wm.update(body: body)
+        destroy_message
+      else
+        storage = ENV['WHERE_FAILURES'].try(:dup) || ''
+        storage << wm.to_json
+        ENV['WHERE_FAILURES'] = storage
```

Comment:
I think it's weird that this works, but
```
(ENV['WHERE_FAILURES'] ||= '') << wm.to_json
```
raises frozen string and symbol into string errors.
Also this is kind of bad, because the resulting json won't be parsable. 🤔 Maybe I can use JSON module...